src/mwindow.c

```diff
@@ -268,7 +272,8 @@ static git_mwindow *new_window(
 
 	ctl->mapped += (size_t)len;
 
-	while (git_mwindow__mapped_limit < ctl->mapped &&
+	while ((git_mwindow__mapped_limit < ctl->mapped ||
+	       git_mwindow__open_limit < ctl->open_windows) &&
```

Comment:
Hum. I think this doesn't exactly work the way we need it to work.Thing is we track mwindows via a list ordered by recency of last usage. If we just call `git_mwindow_close_lru` until using less than `open_limit` files, then we will most likely close mwindows of other files, which doesn't help us at all in that case. Instead, we'd need to have some kind of heuristic that checks which file has the least frequently used mwindows and close all mwindows of that file only. Rinse and repeat until we are below `open_limit`.

Basically, instead of iterating over `git_mwindow`s, we need to iterate over `git_mwindow_file`s and determine which one of them to close.

Reply:
I see, I think.

From a quick look, that surgery might require more time than I really wanted to dedicate right now.

Do you think there is anyone who already knows this code that might be able to work on a proper fix? (If not, I may have to, since this issue is critical for me.)

Reply:
i may be able to help, since it's also affecting us. let me see if i can wrangle a few hours this weekend.

Reply:
I could wrangle a few hours! #5396

Reply:
Hooray! Thanks!! I've been under water fighting another bug, but I'll patch in your PR and give it a whirl soon.